core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala

```diff
@@ -74,12 +80,22 @@ object PythonRunner {
     // Launch Python process
     val builder = new ProcessBuilder((Seq(pythonExec, formattedPythonFile) ++ otherArgs).asJava)
     val env = builder.environment()
+    if (sparkConf.getOption("spark.remote").nonEmpty) {
+      // For non-local remote, pass configurations to environment variables so
+      // Spark Connect client sets them. For local remotes, they will be set
+      // via Py4J.
```

Comment:
Why still take a dependency on Py4j in connect mode?

Reply:
For `bin/pyspark --remote local`, it still launches Spark Connect server via Py4J. This should be fine because the Py4J jars and other dependencies should be present to support `bin/pyspark --remote local` run.

Pure Python packaged ones would not support `bin/pyspark --remote local` and won't need Py4J.